crates/compat/src/http.rs

```diff
@@ -23,6 +26,87 @@ pub struct Request {
     pub body: Vec<u8>,
 }
 
+impl Request {
+    pub async fn from_http(
+        request: http::Request<impl AsyncRead + Unpin>,
+    ) -> anyhow::Result<Request> {
+        let (
+            http::request::Parts {
+                method,
+                uri,
+                headers,
+                ..
+            },
+            mut body,
+        ) = request.into_parts();
+        let content_length = headers
+            .get(CONTENT_LENGTH)
+            .and_then(|content_length| content_length.to_str().ok())
+            .and_then(|content_length| content_length.parse().ok());
+        let header = headers
+            .into_iter()
+            .map(|(name, value)| {
+                let name = name.context("name missing")?;
+                let value = value
+                    .to_str()
+                    .with_context(|| format!("failed to parse `{name}` header value as string"))?;
+                Ok((name.as_str().into(), vec![value.into()]))
+            })
+            .collect::<anyhow::Result<_>>()
+            .context("failed to process request headers")?;
+        let body = if let Some(content_length) = content_length {
+            let mut buf = Vec::with_capacity(content_length);
+            body.take(content_length.try_into().unwrap_or(u64::MAX))
+                .read_to_end(&mut buf)
+                .await
+                .context("failed to read request body")?;
+            buf
+        } else {
+            let mut buf = vec![];
+            body.read_to_end(&mut buf)
+                .await
+                .context("failed to read request body")?;
+            buf
+        };
+        Ok(Request {
+            method: method.as_str().into(),
+            path: uri.path().into(),
+            query_string: uri.query().map(Into::into).unwrap_or_default(),
+            header,
+            body,
+        })
+    }
+}
+
+impl TryFrom<Request> for http::Request<Vec<u8>> {
+    type Error = anyhow::Error;
+
+    fn try_from(
+        Request {
+            method,
+            path,
+            query_string,
+            header,
+            body,
+        }: Request,
+    ) -> Result<Self, Self::Error> {
+        let req = http::Request::builder().method(method.as_str());
+        let req = header
+            .into_iter()
+            .filter_map(|(name, mut values)| {
+                let value = values.pop()?;
+                Some((name, value))
+            })
+            .fold(req, |req, (name, value)| req.header(name, value));
+        let req = if !path.is_empty() || !query_string.is_empty() {
+            req.uri(format!("{path}{query_string}"))
```

Comment:
If path is empty, is this valid?

Reply:
It may be - there's nothing else we can really do here.
If the resulting URI is not valid, then the request construction below will fail, so I don't think this really matters